tree-sitter/src/TreeSitter/GenerateSyntax.hs

```diff
@@ -50,82 +50,93 @@ astDeclarationsForLanguage language filePath = do
 syntaxDatatype :: Ptr TS.Language -> Datatype -> Q [Dec]
 syntaxDatatype language datatype = case datatype of
   SumType (DatatypeName datatypeName) _ subtypes -> do
-    cons <- traverse (constructorForSumChoice datatypeName) subtypes
-    result <- symbolMatchingInstanceForSums language name subtypes
-    pure $ generatedDatatype name cons:result
+    typeParameterName <- newName "a"
+    cons <- traverse (constructorForSumChoice datatypeName typeParameterName) subtypes
+    result <- symbolMatchingInstanceForSums language name subtypes typeParameterName
+    pure $ generatedDatatype name cons typeParameterName:result
   ProductType (DatatypeName datatypeName) _ children fields -> do
-    con <- ctorForProductType datatypeName children fields
-    result <- symbolMatchingInstance language name datatypeName
-    pure $ generatedDatatype name [con]:result
+    typeParameterName <- newName "a"
+    con <- ctorForProductType datatypeName typeParameterName children fields
```

Comment:
If we factored out the bit of `ctorForProductType` which constructs the list of field name/type pairs, could this then be changed to use `ctorForTypes`? What about the use site of `ctorForLeafType`, below?